src/wallet/wallet2.cpp

```diff
@@ -560,6 +556,133 @@ void wallet2::set_seed_language(const std::string &language)
+  rct::addKeys(D_rct, rct::pk2rct(keys.m_account_address.m_spend_public_key), rct::pk2rct(M));  // could have defined add_public_key() under src/crypto
+  crypto::public_key D = rct::rct2pk(D_rct);
+
+  return D;
+}
+//----------------------------------------------------------------------------------------------------
+std::string wallet2::get_subaddress_as_str(const cryptonote::subaddress_index& index) const
+{
+  cryptonote::account_public_address address = get_subaddress(index);
+  return cryptonote::get_account_address_as_str(m_testnet, !index.is_zero(), address);
+}
+//----------------------------------------------------------------------------------------------------
+std::string wallet2::get_integrated_subaddress_as_str(const cryptonote::subaddress_index& index, const crypto::hash8& payment_id) const
+{
+  cryptonote::account_public_address address = get_subaddress(index);
+  return cryptonote::get_account_integrated_address_as_str(m_testnet, !index.is_zero(), address, payment_id);
+}
+//----------------------------------------------------------------------------------------------------
+void wallet2::switch_subaddress_account(uint32_t index_major)
+{
+  if (index_major >= get_num_subaddress_accounts())
+    throw std::runtime_error("index_major is out of bound");
+  m_current_subaddress_account = index_major;
+}
+//----------------------------------------------------------------------------------------------------
+void wallet2::add_subaddress_account(const std::string& label)
+{
+  expand_subaddresses({(uint32_t)get_num_subaddress_accounts(), 0});
+  m_current_subaddress_account = get_num_subaddress_accounts() - 1;
+  m_subaddress_labels[m_current_subaddress_account][0] = label;
+}
+//----------------------------------------------------------------------------------------------------
+void wallet2::add_subaddress(const std::string& label)
+{
+  expand_subaddresses(get_num_subaddresses());
+  m_subaddress_labels[m_current_subaddress_account][get_num_subaddresses() - 1] = label;
+}
+//----------------------------------------------------------------------------------------------------
+void wallet2::expand_subaddresses(const cryptonote::subaddress_index& index)
+{
+  if (m_subaddress_labels.size() <= index.major)
+  {
+    // add new accounts
+    cryptonote::subaddress_index index2;
+    for (index2.major = m_subaddress_labels.size(); index2.major < index.major + SUBADDRESS_LOOKAHEAD_MAJOR; ++index2.major)
+    {
+      for (index2.minor = 0; index2.minor < (index2.major == index.major ? index.minor : 0) + SUBADDRESS_LOOKAHEAD_MINOR; ++index2.minor)
+      {
+        if (m_subaddresses_inv.count(index2) == 0)
+        {
+          crypto::public_key D = get_subaddress_spend_public_key(index2);
+          m_subaddresses[D] = index2;
+          m_subaddresses_inv[index2] = D;
+        }
+      }
+    }
+    m_subaddress_labels.resize(index.major + 1, {""});
+  }
+  else if (m_subaddress_labels[index.major].size() <= index.minor)
```

Comment:
I'm confused. What problem do you see?

> If not, you can't create a new subaddress with an arbitrary amount of proxy addresses.

Can you give me an example scenario?